File: LeetCode-September-Challenge/BagofTokens.js
var bagOfTokensScore = function (tokens, power) {
    tokens.sort((a, b) => a - b); let  i = 0, j = tokens.length - 1, score = 0, res = 0;
    while(i <= j){
        if(tokens[i] <= power){
            score++;
            power -= tokens[i++];
            if(res < score) res = score;
        } else if(score > 0){
            score--;
            power += tokens[j--];
        } else break
    }
    return res;
};
let tokens = [81, 45, 8], power = 32;
console.log(bagOfTokensScore(tokens, power));

File: Year-2022/LeetCode-September-Challenge/BagofTokens.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        sort(tokens.begin(), tokens.end()); int i = 0, j = tokens.size() - 1, score = 0, res = 0;
        while(i <= j){
            if(tokens[i] <= power){
                score++;
                power -= tokens[i++];
                if(score > res) res = score;
            } else if(score > 0){
                score--;
                power += tokens[j--];
            } else break;
        }
        return  res;
    }
};
int main(){
    vector<int> tokens = {100};
    int power = 50;
    Solution s; cout<<s.bagOfTokensScore(tokens, power)<<endl;
    return 0;
}